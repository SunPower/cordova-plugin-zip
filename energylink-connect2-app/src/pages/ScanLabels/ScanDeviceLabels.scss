.scan {
  #scandit {
    min-height: 300px;
  }
}
