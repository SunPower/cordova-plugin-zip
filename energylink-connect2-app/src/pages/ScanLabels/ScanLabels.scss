.scan-labels {
  display: grid;
  grid-template-rows: minmax(20px, 30px) 1fr minmax(30px, 60px) minmax(
      10px,
      20px
    ) minmax(30px, 60px) minmax(20px, 40px);
  height: 100%;
  align-items: center;
  justify-content: center;

  .has-no-border {
    border: none;
  }

  .trigger-scan {
    width: fit-content;
    justify-self: center;
  }

  #scandit {
    height: 100%;
    width: 105%;
    margin-left: -10px;
  }
}
