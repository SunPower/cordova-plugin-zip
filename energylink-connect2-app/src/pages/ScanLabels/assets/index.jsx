export { ReactComponent as BarcodeIcon } from './Barcode.svg'
