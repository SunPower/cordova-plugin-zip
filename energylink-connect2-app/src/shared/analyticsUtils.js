import {
  __,
  compose,
  curry,
  gt,
  isEmpty,
  join,
  lensIndex,
  map,
  not,
  over,
  props,
  propSatisfies,
  toLower,
  toUpper,
  when
} from 'ramda'

export const isNotEmpty = compose(not, isEmpty)

export const capitalizeWord = when(
  propSatisfies(gt(__, 0), 'length'),
  compose(join(''), over(lensIndex(0), toUpper), toLower)
)

const stringValueOrEmptyString = stringValue => stringValue || ''

const getStringProps = curry(compose(map(stringValueOrEmptyString), props))

export const getUserProfile = getStringProps([
  'uniqueId',
  'firstName',
  'lastName',
  'email',
  'dealerName',
  'recordType'
])
