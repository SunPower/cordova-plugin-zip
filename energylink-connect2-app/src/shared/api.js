import Swagger from 'swagger-client'
let apiParty, apiSite, apiFirmware, apiScanSN

const requestOptions = access_token => ({
  requestInterceptor: req => {
    if (access_token) req.headers['Authorization'] = `Bearer ${access_token}`
    return req
  }
})

export async function getApiFirmware() {
  if (!apiFirmware)
    apiFirmware = await Swagger(process.env.REACT_APP_SWAGGER_FIRMWARE)
  return apiFirmware
}
export async function getApiPVS() {
  return await Swagger(process.env.REACT_APP_PVS_ADDRESS)
}
export async function getApiDevice(access_token) {
  return await Swagger(
    process.env.REACT_APP_SWAGGER_DEVICE,
    requestOptions(access_token)
  )
}
export async function getApiParty(access_token) {
  apiParty = await Swagger(
    process.env.REACT_APP_SWAGGER_PARTY,
    requestOptions(access_token)
  )
  return apiParty
}
export async function getApiSearch(access_token) {
  return await Swagger(
    process.env.REACT_APP_SWAGGER_SEARCH,
    requestOptions(access_token)
  )
}
export async function getApiSite(access_token) {
  if (!apiSite)
    apiSite = await Swagger(
      process.env.REACT_APP_SWAGGER_SITE,
      requestOptions(access_token)
    )
  return apiSite
}
export async function getApiScanSN() {
  if (!apiScanSN)
    apiScanSN = await Swagger(process.env.REACT_APP_SCAN_SERIAL_NUMBERS)
  return apiScanSN
}
export async function getApiAuth(access_token) {
  return await Swagger(
    process.env.REACT_APP_SWAGGER_AUTH,
    requestOptions(access_token)
  )
}
export const storageSwaggerTag = 'Commissioning'
