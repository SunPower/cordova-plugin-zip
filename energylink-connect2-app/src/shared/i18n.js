import { useMemo } from 'react'
import { useSelector } from 'react-redux'

import * as languages from './locales'
import * as localizedComponents from './localizedComponents'

const getTranslate = translations => (key, ...params) => {
  let value = translations[key] || key
  if (params) {
    params.forEach((p, idx) => {
      value = value.replace(`{${idx}}`, p)
    })
  }
  return value
}

export const useI18n = () => {
  const language = useSelector(state => state.language)
  const { locale = 'en' } = language
  return useMemo(() => getTranslate(languages[locale]), [locale])
}

export const useI18nComponent = () => {
  const language = useSelector(state => state.language)
  const { locale = 'en' } = language || {}

  return useMemo(
    () => componentName => {
      const path = `${locale}_${componentName}`
      const Component = localizedComponents[path] || {}
      if (typeof Component !== 'function') {
        throw new Error(`Component not found: ${path}`)
      }
      return Component()
    },
    [locale]
  )
}
