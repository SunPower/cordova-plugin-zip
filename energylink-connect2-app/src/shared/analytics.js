import appVersion from '../macros/appVersion.macro'
import { capitalizeWord, getUserProfile } from 'shared/analyticsUtils'
import { MIXPANEL_EVENT_QUEUED } from 'state/actions/analytics'
import { getAppleDeviceFamily } from 'shared/appleDevicesTable'
import { prop } from 'ramda'

const getAppAndDeviceProperties = () => {
  const { device } = window
  const { model, version, platform, manufacturer } = device

  let deviceModel
  let deviceBrowserVersion

  if (platform === 'iOS') {
    deviceModel = getAppleDeviceFamily(model)
    deviceBrowserVersion = version
  } else {
    deviceModel = `${manufacturer}, ${model}`
  }

  return {
    'App Build': appVersion(),
    'App Flavor': process.env.REACT_APP_FLAVOR,
    'OS Version': version,
    ...(deviceBrowserVersion ? { $browser_version: deviceBrowserVersion } : {}),
    $device: deviceModel
  }
}

export const loggedIn = user => {
  const { mixpanel } = window
  const {
    uniqueId: userId,
    firstName,
    lastName,
    email,
    dealerName,
    recordType: dealerType
  } = getUserProfile(user)
  const appAndDeviceProperties = getAppAndDeviceProperties()

  mixpanel.register(appAndDeviceProperties)
  mixpanel.identify(userId)
  mixpanel.people.set(appAndDeviceProperties)
  mixpanel.people.set({
    $first_name: firstName,
    $last_name: lastName,
    $email: email,
    'Dealer Name': dealerName,
    'Dealer Type': capitalizeWord(dealerType)
  })
  mixpanel.track('Login', { Success: true })

  return MIXPANEL_EVENT_QUEUED('Login - Success')
}

export const loginFailed = () => {
  const { mixpanel } = window
  const appAndDeviceProperties = getAppAndDeviceProperties()

  mixpanel.register(appAndDeviceProperties)
  mixpanel.track('Login', { Success: false })

  return MIXPANEL_EVENT_QUEUED('Login - Failed')
}

export const scanPVS = ({ pvsSN, entryMethod }, event) => {
  const { mixpanel } = window

  mixpanel.register({ 'PVS SN': pvsSN })
  mixpanel.track('Scan PVS Tag', {
    'Entry Method': entryMethod
  })

  return MIXPANEL_EVENT_QUEUED(event)
}

export const siteNotFound = ({
  uniqueId: userId,
  firstName,
  lastName,
  email,
  dealerName,
  recordType: dealerType
}) => {
  const { mixpanel } = window
  const appAndDeviceProperties = getAppAndDeviceProperties()

  mixpanel.register(appAndDeviceProperties)
  mixpanel.identify(userId)
  mixpanel.people.set(appAndDeviceProperties)
  mixpanel.people.set({
    $first_name: firstName,
    $last_name: lastName,
    $email: email,
    'Dealer Name': dealerName,
    'Dealer Type': capitalizeWord(dealerType)
  })
  mixpanel.track('Find Site', { Found: false })

  return MIXPANEL_EVENT_QUEUED('Find Site - site not found')
}

export const siteFound = (
  {
    uniqueId: userId,
    firstName,
    lastName,
    email,
    dealerName,
    recordType: dealerType
  },
  { city, st_id, postalCode, siteKey, commissioned }
) => {
  const { mixpanel } = window
  const appAndDeviceProperties = getAppAndDeviceProperties()

  mixpanel.register(appAndDeviceProperties)
  mixpanel.identify(userId)
  mixpanel.people.set(appAndDeviceProperties)
  mixpanel.people.set({
    $first_name: firstName,
    $last_name: lastName,
    $email: email,
    'Dealer Name': dealerName,
    'Dealer Type': capitalizeWord(dealerType)
  })
  mixpanel.track('Find Site', {
    Found: true,
    ...{
      $city: city,
      State: st_id,
      'Zip Code': postalCode,
      'Site ID': siteKey,
      Commissioned: commissioned
    }
  })

  return MIXPANEL_EVENT_QUEUED('Find Site - site found')
}

export const saveConfiguration = config => {
  const { mixpanel } = window
  mixpanel.track('Configure', config)
  return MIXPANEL_EVENT_QUEUED('Configure - submit site config')
}

export const saveInventory = (inventory, hasEss) => {
  const { mixpanel } = window

  mixpanel.register({
    'Equinox Storage Site': hasEss
  })

  mixpanel.track('Inventory', {
    'AC Module Count': prop('AC_MODULES', inventory),
    'ESS System Size': hasEss ? prop('ESS', inventory) : 'None',
    'Equinox Storage Site': hasEss
  })
  return MIXPANEL_EVENT_QUEUED('Inventory')
}

export const deviceResume = pageResume => {
  const { mixpanel } = window
  mixpanel.register({
    Page: pageResume
  })
  return MIXPANEL_EVENT_QUEUED('User opens app - Register page resume')
}

/**
 *
 * @param {string}[interfaceType] - The type of connection the user has. either can be  PVS, Cell, or Customer WiFi
 * @return {Action<null, null>}
 */
export const internetConnection = (interfaceType = '') => {
  const { mixpanel } = window
  mixpanel.register({
    'Internet Connection': interfaceType
  })
  return MIXPANEL_EVENT_QUEUED('User opens app - Register page resume')
}

export const firmwareUpdateEvent = data => {
  const { mixpanel } = window
  mixpanel.track('Firmware Update', {
    'Current FW Version': prop('fromFWVersion', data),
    'Current BN Version': prop('fromBuildNumber', data),
    'New FW Version': prop('toFWVersion', data),
    'New BN Version': prop('toBuildNumber', data),
    Success: prop('success', data)
  })
  return MIXPANEL_EVENT_QUEUED('Firmware update - Update successful')
}

export const timeMixPanelEvent = name => {
  if (!name) throw new Error('The name parameter is required')
  const { mixpanel } = window
  mixpanel.time_event(name)
}

export const commissionSite = ({ duration, timeConfiguring }) => {
  const { mixpanel } = window
  mixpanel.track('Commission Site', {
    $duration: duration,
    'Time Configuring': timeConfiguring
  })
  return MIXPANEL_EVENT_QUEUED(
    'Commission Site - first successful site configuration'
  )
}
