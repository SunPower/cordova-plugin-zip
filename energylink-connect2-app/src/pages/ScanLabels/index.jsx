import React, { useEffect, useRef } from 'react'
import { compose, map } from 'ramda'
import { useDispatch, useSelector } from 'react-redux'
import { useHistory } from 'react-router-dom'
import { useI18n } from 'shared/i18n'
import { scanM } from 'shared/scandit'
import { buildSN } from 'shared/utils'
import { ADD_PVS_SN } from 'state/actions/pvs'
import paths from 'routes/paths'

import './ScanLabels.scss'

const addSN = dispatch => compose(dispatch, ADD_PVS_SN, buildSN)
const addCodes = compose(map, addSN)
const scanMatrix = compose(scanM, addCodes)
const startScanning = dispatch => {
  if (window.Scandit) return scanMatrix(dispatch)
}

function ScanDeviceLabels() {
  const t = useI18n()
  const dispatch = useDispatch()
  const history = useHistory()
  const { serialNumbers } = useSelector(state => state.pvs)

  const onDone = useRef(null)

  const finishedScanning = () => {
    if (typeof onDone.current === 'function') {
      onDone.current()
      history.push(paths.PROTECTED.SN_LIST.path)
    }
  }

  const triggerManualEntry = () => {
    history.push({
      pathname: paths.PROTECTED.SN_LIST.path,
      state: {
        isManualModeDefault: true
      }
    })
  }

  useEffect(() => {
    onDone.current = startScanning(dispatch)
  }, [dispatch])

  return (
    <div className="scan-labels is-vertical has-text-centered pl-10 pr-10">
      <span className="is-uppercase has-text-weight-bold mb-10">
        {t('SCAN_EQUIPMENT')}
      </span>

      <div id="scandit" />

      <div className="hint-text mt-15 pl-15 pr-15">{t('BULK_SCAN_HINT')}</div>

      <div className="has-text-white mt-10">
        {t('FOUND_SN', serialNumbers.length)}
      </div>

      <button
        className="button is-primary trigger-scan mt-15"
        onClick={finishedScanning}
      >
        {t('DONE')}
      </button>

      <button
        className="button has-text-centered is-uppercase is-secondary has-no-border"
        onClick={triggerManualEntry}
      >
        {t('SN_MANUAL_ENTRY')}
      </button>
    </div>
  )
}

export default ScanDeviceLabels
