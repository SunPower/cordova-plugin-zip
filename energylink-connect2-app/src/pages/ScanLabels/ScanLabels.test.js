import React from 'react'
import ScanLabels from '.'
import * as i18n from 'shared/i18n'

describe('Bulk Scan screen', () => {
  beforeEach(() => {
    jest
      .spyOn(i18n, 'useI18n')
      .mockImplementation(path => (key, ...params) =>
        `${key.toUpperCase()} ${params.join('_')}`.trim()
      )
  })

  test('renders correctly', () => {
    const { component } = mountWithProvider(<ScanLabels />)({
      pvs: {
        serialNumbers: [],
        fetchingSN: false
      }
    })
    expect(component).toMatchSnapshot()
  })
})
