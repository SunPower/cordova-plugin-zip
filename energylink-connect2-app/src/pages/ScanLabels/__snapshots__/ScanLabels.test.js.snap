// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bulk Scan screen renders correctly 1`] = `
<BrowserRouter
  keyLength={0}
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <ScanDeviceLabels>
        <div
          className="scan-labels is-vertical has-text-centered pl-10 pr-10"
        >
          <span
            className="is-uppercase has-text-weight-bold mb-10"
          >
            SCAN_EQUIPMENT
          </span>
          <div
            id="scandit"
          />
          <div
            className="hint-text mt-15 pl-15 pr-15"
          >
            BULK_SCAN_HINT
          </div>
          <div
            className="has-text-white mt-10"
          >
            FOUND_SN 0
          </div>
          <button
            className="button is-primary trigger-scan mt-15"
            onClick={[Function]}
          >
            DONE
          </button>
          <button
            className="button has-text-centered is-uppercase is-secondary has-no-border"
            onClick={[Function]}
          >
            SN_MANUAL_ENTRY
          </button>
        </div>
      </ScanDeviceLabels>
    </Provider>
  </Router>
</BrowserRouter>
`;
